### PR TITLE
Cswang/input

### PR DESCRIPTION
**Pull request by ttw225**
Finish Restful API for 1. receive data and send to IOTA 2. search the transaction data from IOTA.

### CODE REVIEW

File: api/bems.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Client
+import socket
+import sys
+import requests
+import json
+
+json_data = {
```
Comment (grandq33769) [inline api/bems.py]: Would you please write the sample data as a file ?

File: api/docker.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+docker build --no-cache -t api_docker:latest . &
```
Comment (grandq33769) [inline api/docker.sh]: You may try `makefile` next time.

File: api/resources/get_transaction.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+from flask_restful import Resource
+from flask import request
+from config import app
+import iota
+import json
+
+
+class get_transaction (Resource):
```
Comment (grandq33769) [inline api/resources/get_transaction.py]: Python class should be uppercase in the first letter.

File: api/tangle.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    pt.append(iota.ProposedTransaction(address=iota.Address(app.config['TARGETADDRESS']),  # 81 trytes long address
+                                       message=iota.TryteString.from_unicode(
+        send_data),
+        # Up to 27 trytes
+        tag=iota.Tag(b'RBTC9D9DCDEAKDCDFD9DSC'),
+        value=0))
+
+    api = iota.Iota(app.config['API_URI'])
+
+    print("[INFO] Preparing/Broadcasting... Wait please...")
```
Comment (grandq33769) [inline api/tangle.py]: You may try `logging` next time.

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1,31 @@
+-i https://pypi.org/simple
```
Comment (grandq33769) [inline requirements.txt]: Why there are `requirement.txt` in root directory and `./api` ?

File: database/test_performance.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
+import pymysql
```
Comment (grandq33769) [inline database/test_performance.py:1]: You may try `pytest` next time.

File: README.md [hunk 7]
```diff
@@ -1,4 +1,123 @@
...
+        request reject by Tangle
+
+# Run
+Pull from docker hub
+
+`docker pull ttw225/api_docker`
+
+Run in background
+
+`docker run -p 5000:5000 ttw225/api_docker:latest &`
```
Comment (grandq33769) [inline README.md]: You may use `docker run -d`

File: api/cems.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import socket
+# import iota
+import datetime
+import requests
+import json
+import hashlib
+import base64
+
+
+def get_data(Tx, data):
```
Comment (grandq33769) [inline api/cems.py]: Do you think is it possible to let this function to become `utils` func ?

File: api/cems.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    hash_data = hashlib.sha256(
+        json_data).hexdigest().encode()
+    base64_data = base64.b64encode(hash_data).decode()
+    if base64_data == message['message']['value']:
+        return True
+    else:
+        return False
+
+
+# Socket
```
Comment (grandq33769) [inline api/cems.py:12]: How you start an https server while accepting socket?

**Review by grandq33769 — COMMENTED** [review]
Maybe you may use `pylint` next time.
Considering how to run an http server while accepting socket.

File: api/docker.sh (hunk 2 above)
Comment (ttw225) [reply api/docker.sh]: Create a Makefile and delete docker.sh.

File: api/resources/get_transaction.py (hunk 3 above)
Comment (ttw225) [reply api/resources/get_transaction.py]: Fix in [7e1fd76](https://github.com/NCKU-CCS/energy-blockchain/pull/2/commits/7e1fd7630ef164a27989567a099cb5d3971ccfec).

File: README.md (hunk 7 above)
Comment (ttw225) [reply README.md]: Fix in ba15c60.

File: requirements.txt (hunk 5 above)
Comment (ttw225) [reply requirements.txt]: Actually it is a file for testing.
Also have `Dockerfile` in both root directory and `./api`.
Delete `Dockerfile` in root and delete `requirement.txt` in root.
Fix in ad37afa.

File: api/tangle.py (hunk 4 above)
Comment (ttw225) [reply api/tangle.py]: Fix in bc1891b.

File: api/cems.py (hunk 8 above)
Comment (ttw225) [reply api/cems.py]: This function is used to check bems's data's hash is same as data on IOTA or not.
Who's `utils` function ? PyOTA's?

File: api/tangle.py (hunk 4 above)
Comment (ttw225) [reply api/tangle.py]: Update at 973aa5c.

File: api/bems.py (hunk 1 above)
Comment (ttw225) [reply api/bems.py]: Fix in 9a49133.

File: api/cems.py (hunk 8 above)
Comment (ttw225) [reply api/cems.py]: Yes, it is.

Comment (grandq33769) [reply api/cems.py]: so may you remove this function in this .py ?

File: api/resources/hems.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+from flask import request
+from models.schema.hems import hems
+from config import app
+import json
+import hashlib
+import base64
+from tangle import send_to_iota
+
+
+class Hems (Resource):
```
Comment (grandq33769) [inline api/resources/hems.py]: The requirement has changed to `bems` & `cems` , not `bems` and `hems` .

File: api/tangle.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    # preparing transactions
+    pt = []
+
+    starttime = datetime.now()
+    logging.info("START")
+    pt.append(iota.ProposedTransaction(address=iota.Address(app.config['TARGETADDRESS']),  # 81 trytes long address
+                                       message=iota.TryteString.from_unicode(
+        send_data),
+        # Up to 27 trytes
+        tag=iota.Tag(b'RBTC9D9DCDEAKDCDFD9DSC'),
```
Comment (grandq33769) [inline api/tangle.py:19]: May you change it to a configuration parameter?

File: api/resources/get_transaction.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
+from flask_restful import Resource
+from flask import request
+from config import app
+import iota
+import json
+
+
+class Get_transaction (Resource):
+    def get(self, name):
```
Comment (grandq33769) [inline api/resources/get_transaction.py]: Do you think `tx_hash` is better than `name` ?

File: api/models/schema/hems.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
+from common.ma import ma
+from marshmallow import validate
+
+# HEMS
+
+
+class hems():
```
Comment (grandq33769) [inline api/models/schema/hems.py]: We should re-consider the situation of dr event.

File: api/bems.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
+# Client
+import socket
+import sys
+import requests
+import json
+from utils.logging import logging
+with open('dr.json', 'r') as f:
+    json_data = json.load(f)
+
+r = requests.post(
```
Comment (grandq33769) [inline api/bems.py:12]: Could you please change the url as a configuration parameter?

**Review by grandq33769 — COMMENTED** [review]
We mainly need to re-think the situation of dr event.

File: api/cems.py (hunk 8 above)
Comment (ttw225) [reply api/cems.py]: Resolve in 4c34f06.

File: api/resources/hems.py (hunk 10 above)
Comment (ttw225) [reply api/resources/hems.py]: I think the resolve way is: move `send_dr_event` from hems to bems, and remove hems.
Is this correct?

File: api/tangle.py (hunk 11 above)
Comment (ttw225) [reply api/tangle.py:19]: Did you mean `address` and `tag`?

File: api/resources/get_transaction.py (hunk 12 above)
Comment (ttw225) [reply api/resources/get_transaction.py]: Fix in 629e94e.

File: api/bems.py (hunk 14 above)
Comment (ttw225) [reply api/bems.py:12]: Fix in 1667759.

File: api/main.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
+from flask import Flask
+from flask_restful import Api
+from resources.hems import Hems
```
Comment (grandq33769) [inline api/main.py]: Switch from `Hems` to `Cems` ?

File: api/utils/utils.py [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+    import requests
+    import json
+    import hashlib
+    import base64
+    r = requests.get(
+        'http://localhost:5000/get_transaction/'+Tx)
+    message = r.json()
+    json_data = json.dumps(data).encode('utf-8')
+    hash_data = hashlib.sha256(
+        json_data).hexdigest().encode()
```
Comment (grandq33769) [inline api/utils/utils.py]: Why there is a newline? you can use `flake8` or `yapf` to format the code

File: api/utils/utils.py [hunk 17]
```diff
@@ -0,0 +1,16 @@
+def verify_data(Tx, data):
+    import requests
```
Comment (grandq33769) [inline api/utils/utils.py]: Please move all import to the top.

File: api/utils/utils.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+    import json
+    import hashlib
+    import base64
+    r = requests.get(
+        'http://localhost:5000/get_transaction/'+Tx)
```
Comment (grandq33769) [inline api/utils/utils.py]: Cloud you please change the `http://localhost:5000` to configuration parameter?

File: api/resources/get_transaction.py [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class Get_transaction (Resource):
+    def get(self, tx_hash):
+        if len(tx_hash) != 81:
+            return {
+                'message': 'Type Not Included!'
+            }, 400
+
+        api = iota.Iota(app.config['API_URI'])
```
Comment (grandq33769) [inline api/resources/get_transaction.py]: `get_transaction` can be a utils function.

**Review by grandq33769 — COMMENTED** [review]
1. Switch all `hems` module to `cems`
2. Format the code
3. Create `get_transaction` utils function

File: api/utils/utils.py (hunk 16 above)
Comment (ttw225) [reply api/utils/utils.py]: VS code auto format

File: api/main.py (hunk 15 above)
Comment (ttw225) [reply api/main.py]: Fix in 0f8e586.

File: api/resources/hems.py (hunk 10 above)
Comment (ttw225) [reply api/resources/hems.py]: hems->cems
Fix in 0f8e586.

File: api/resources/get_transaction.py (hunk 19 above)
Comment (ttw225) [reply api/resources/get_transaction.py]: Fix in f01cf8f.

File: api/utils/utils.py (hunk 17 above)
Comment (ttw225) [reply api/utils/utils.py]: Fix in f01cf8f.

File: api/utils/utils.py (hunk 16 above)
Comment (ttw225) [reply api/utils/utils.py]: Fix in f01cf8f.

File: api/utils/utils.py (hunk 18 above)
Comment (ttw225) [reply api/utils/utils.py]: Fix in a379c2d.

File: api/tangle.py (hunk 11 above)
Comment (ttw225) [reply api/tangle.py:19]: Conclusion is no need to change.

File: api/resources/verify.py [hunk 20]
```diff
@@ -0,0 +1,17 @@
+import requests
```
Comment (grandq33769) [inline api/resources/verify.py]: Should this script locate in `api/utils` ?

File: api/utils/get_transaction.py [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+        if len(tx_hash) != 81:
+            return {
+                'message': 'Type Not Included!'
+            }, 400
+
+        api = iota.Iota(app.config['API_URI'])
+
+        Tx = tx_hash
+        try:
+            data = api.get_trytes(hashes=[Tx])
```
Comment (grandq33769) [inline api/utils/get_transaction.py]: From line 19 to 27, this may be a utility function and put it in `api/utils` ?

**Review by grandq33769 — COMMENTED** [review]
Please review the comment and fix it if you can. Thank you !
After resolving all these problems, I think this merge request can be approved.

File: api/resources/verify.py (hunk 20 above)
Comment (ttw225) [reply api/resources/verify.py]: Fix in f71aa2f.

File: api/utils/get_transaction.py (hunk 21 above)
Comment (ttw225) [reply api/utils/get_transaction.py]: Fix in f71aa2f.

**Comment by grandq33769** [comment]
Awesome. LGTM

**Pull request merged**